central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,624 @@
+	deploymentDatastore deploymentDS.DataStore, watchedImageDatastore watchedImageDS.DataStore,
+	collectionDatastore collectionDS.DataStore, collectionQueryRes collectionDS.QueryResolver,
+	notificationProcessor notifier.Processor, cronScheduler *cron.Cron, schema *graphql.Schema) *scheduler {
+	s := &scheduler{
+		reportConfigToEntryIDs:  make(map[string]cron.EntryID),
+		reportConfigDatastore:   reportConfigDatastore,
+		reportMetadataStore:     reportMetadataStore,
+		reportSnapshotStore:     reportSnapshotStore,
+		notifierDatastore:       notifierDatastore,
+		deploymentDatastore:     deploymentDatastore,
+		watchedImageDatastore:   watchedImageDatastore,
+		collectionDatastore:     collectionDatastore,
+		collectionQueryResolver: collectionQueryRes,
+		notificationProcessor:   notificationProcessor,
+		reportsQueue:            make([]*ReportRequest, 0),
+		sendNewReports:          make(chan bool, 200),
+		runningReportConfigs:    set.NewStringSet(),
+		Schema:                  schema,
+
+		stopper:         concurrency.NewStopper(),
+		cron:            cronScheduler,
+		concurrencySema: semaphore.NewWeighted(int64(env.ReportExecutionMaxConcurrency.IntegerSetting())),
+	}
+	return s
+}
+
+/* Concurrency and scheduling functions */
+
+func (s *scheduler) Start() {
+	ctx := sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS, storage.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.WorkflowAdministration)))
+
+	s.queuePendingReports(ctx)
+	s.queueScheduledReports(ctx)
+	go s.runReports()
+	go s.runReports()
+}
+
+func (s *scheduler) Stop() {
+	s.stopper.Client().Stop()
+	err := s.stopper.Client().Stopped().Wait()
+	if err != nil {
+		log.Errorf("Error stopping vulnerability report scheduler : %v", err)
+	}
+}
+
+func (s *scheduler) runReports() {
+	defer s.stopper.Flow().ReportStopped()
+	for {
+		select {
+		case <-s.stopper.Flow().StopRequested():
+			return
+		case <-s.sendNewReports:
+			reportRequest := s.selectNextRunnableReport()
+			if reportRequest == nil {
+				continue
+			}
+			if err := s.concurrencySema.Acquire(scheduledCtx, 1); err != nil {
+				log.Errorf("Error acquiring semaphore to run new report: %v", err)
+				continue
+			}
+			log.Infof("Executing report '%s' at %v", reportRequest.ReportConfig.GetName(), time.Now().Format(time.RFC822))
+			go s.runSingleReport(reportRequest)
+		}
+	}
+}
+
+func (s *scheduler) selectNextRunnableReport() *ReportRequest {
+	s.schedulerLock.Lock()
+	defer s.schedulerLock.Unlock()
+
+	reportIdx := sliceutils.FindMatching(s.reportsQueue, func(req *ReportRequest) bool {
+		return !s.runningReportConfigs.Contains(req.ReportConfig.GetId())
+	})
+
+	if reportIdx < 0 {
+		// did not find any reports to run
+		return nil
+	}
+	request := s.reportsQueue[reportIdx]
+	s.reportsQueue = slices.Delete(s.reportsQueue, reportIdx, reportIdx+1)
+	s.runningReportConfigs.Add(request.ReportConfig.GetId())
```

Comment:
This can be simplified with `ValueStream`.
stream.Next().
If running, push back to the end of the queue
else, add to the running report

Reply:
The reason I did not want to avoid pushing at the end of queue is to be able to process report requests roughly in the order they arrived. If there is a report with same config already running, a request would still be picked up next time. Pushing to the end of queue would mean it a request missed its chance now, it will be picked up at the end.

Reply:
I could use a doubly linked list here too, but that seemed like an overkill for this.

Reply:
I am concerned about the rewrites happening because of the array.

Reply:
Changed to using the doubly linked list.